### PR TITLE
ENH: Add API to select items in the DICOM browser

### PR DESCRIPTION
**Pull request by lassoan**
Allow selecting patient/study/series in the DICOM browser GUI.

This is can be used for example for highlighting new items after DICOM import.

### CODE REVIEW

**Review by pieper — APPROVED** [review]
LGTM 👍

File: Libs/DICOM/Widgets/ctkDICOMTableManager.cpp [hunk 1]
```diff
@@ -251,6 +251,27 @@ QStringList ctkDICOMTableManager::currentSeriesSelection()
   return d->seriesTable->currentSelection();
 }
 
+//------------------------------------------------------------------------------
+void ctkDICOMTableManager::setCurrentPatientsSelection(const QStringList& uids)
+{
+  Q_D(ctkDICOMTableManager);
+  return d->patientsTable->setCurrentSelection(uids);
```
Comment (jcfr) [inline Libs/DICOM/Widgets/ctkDICOMTableManager.cpp:258]:
```suggestion
  d->patientsTable->setCurrentSelection(uids);
```

File: Libs/DICOM/Widgets/ctkDICOMTableManager.cpp [hunk 2]
```diff
@@ -251,6 +251,27 @@ QStringList ctkDICOMTableManager::currentSeriesSelection()
...
+}
+
+//------------------------------------------------------------------------------
+void ctkDICOMTableManager::setCurrentStudiesSelection(const QStringList& uids)
+{
+  Q_D(ctkDICOMTableManager);
+  return d->studiesTable->setCurrentSelection(uids);
```
Comment (jcfr) [inline Libs/DICOM/Widgets/ctkDICOMTableManager.cpp:265]:
```suggestion
  d->studiesTable->setCurrentSelection(uids);
```

File: Libs/DICOM/Widgets/ctkDICOMTableManager.cpp [hunk 3]
```diff
@@ -251,6 +251,27 @@ QStringList ctkDICOMTableManager::currentSeriesSelection()
...
+}
+
+//------------------------------------------------------------------------------
+void ctkDICOMTableManager::setCurrentSeriesSelection(const QStringList& uids)
+{
+  Q_D(ctkDICOMTableManager);
+  return d->seriesTable->setCurrentSelection(uids);
```
Comment (jcfr) [inline Libs/DICOM/Widgets/ctkDICOMTableManager.cpp:272]:
```suggestion
  d->seriesTable->setCurrentSelection(uids);
```

**Review by jcfr — CHANGES_REQUESTED** [review]

**Comment by jcfr** [comment]
Consider rebasing the topic so that we avoid commit like `Merge branch 'master' into add-dicombrowser-select-api` making it hard to cherry-pick or revert changes.

**Comment by pieper** [comment]
@lassoan github reported this was out of date with master so I merged it and there are no conflicts but I wondered why you hadn't started from the master branch?  Maybe better to rebase instead of squash as @jcfr suggests.

**Comment by lassoan** [comment]
> I wondered why you hadn't started from the master branch?

No specific reason, it was just what I had in my build tree.

Thanks for the review and comments, I'll rebase on latest master.

File: Libs/DICOM/Widgets/ctkDICOMBrowser.cpp [hunk 4]
```diff
@@ -1749,3 +1749,52 @@ void ctkDICOMBrowser::onIndexingComplete(int patientsAdded, int studiesAdded, in
...
+  if (level == ctkDICOMModel::PatientType)
+  {
+    d->dicomTableManager->setCurrentPatientsSelection(uids);
+  }
+  else if (level == ctkDICOMModel::StudyType)
+  {
+    // Select parent patient to make sure the requested studies
+    // are listed in the study table
+    QStringList patientUids;
+    for (QString uid : uids)
```
Comment (jcfr) [inline Libs/DICOM/Widgets/ctkDICOMBrowser.cpp:1766]: For correctness:

```suggestion
    for (const QString& uid : uids)
```

https://stackoverflow.com/questions/15176104/range-based-loop-get-item-by-value-or-reference-to-const/15176127#15176127

File: Libs/DICOM/Widgets/ctkDICOMBrowser.cpp [hunk 5]
```diff
@@ -1749,3 +1749,52 @@ void ctkDICOMBrowser::onIndexingComplete(int patientsAdded, int studiesAdded, in
...
+    this->setSelectedItems(ctkDICOMModel::PatientType, patientUids);
+
+    d->dicomTableManager->setCurrentStudiesSelection(uids);
+  }
+  else if (level == ctkDICOMModel::SeriesType)
+  {
+    // Select parent patients and studies to make sure the requested series
+    // are listed in the series table
+    QStringList studyUids;
+    for (QString uid : uids)
```
Comment (jcfr) [inline Libs/DICOM/Widgets/ctkDICOMBrowser.cpp:1783]:
```suggestion
    for (const QString& uid : uids)
```

**Comment by lassoan** [comment]
@jcfr I've updated ctkDICOMTableView::setCurrentSelection to use `d->tblDicomDatabaseView->selectionModel()->isSelected(index)`, because selection mode is "full row", so it is enough to check that the item in the first column is selected.

**Comment by lassoan** [comment]
@jcfr Thanks for your comments, I addressed all of them, let me know if this is good to go now.

**Pull request merged**